### PR TITLE
Refactor  TestDevicePluginReRegistration for reliability/readability

### LINKED ISSUE #55179: The current kubelet device plugin code could fail to remove some devices that are no longer advertised by a re-registered device plugin
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:
Consider this corner case scenario: a device plugin registers with Kubelet for resourceA and lists two devices: dev1 and dev2. Then a new device plugin implementation is deployed that re-registers resourceA with Kubelet but somehow advertises with different device ids, say dev3 and dev4. With the current implementation, we will end up having dev1, dev2, dev3, and dev4 for resourceA incorrectly.

To fix this problem, one possible solution is to pass list of current devices in newEndpoint(). For the first registered device plugin, the list will be empty. For the subsequent re-registered plugin, the list will be what reported by its previous device plugin.

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:
The problem should be able to be reproduced with a unit test.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by lichuqiang**
**What this PR does / why we need it**:
#55179 related unit test update, after #52149, and some refactoring for extensiblity

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #55179 

**Special notes for your reviewer**:
/cc @jiayingz 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubelet/cm/deviceplugin/manager_test.go [hunk 1]
```diff
@@ -56,73 +58,97 @@ func TestDevicePluginReRegistration(t *testing.T) {
...
 
-	// Test the scenario that a plugin re-registers with different devices.
-	p3 := NewDevicePluginStub(devsForRegistration, pluginSocketName+".third")
-	err = p3.Start()
-	require.NoError(t, err)
-	p3.Register(socketName, testResourceName)
-	// Wait for the second callback to be issued.
-	<-callbackChan
+		<-callbackChan
+		require.Equal(t, testCase.expectedDevsLen, maintainedDevices[testResourceName].Len(), "Expected length [%d], actual length [%d]", testCase.expectedDevsLen, maintainedDevices[testResourceName].Len())
```
Comment (ScorpioCPH) [inline pkg/kubelet/cm/deviceplugin/manager_test.go]: As test description `Devices shouldn't change`, should we compare devices' id one by one to much sure re-registration works well, not just compare the length of devices?

**Review by ScorpioCPH — COMMENTED** [review]
HI, just one quick comment.

File: pkg/kubelet/cm/deviceplugin/manager_test.go (hunk 1 above)
Comment (jiayingz) [reply pkg/kubelet/cm/deviceplugin/manager_test.go]: +1. Perhaps it is easier to maintain devs and devsForRegistration as sets.String that has some helper function to compare two sets.String (Equal) or convert to slice (UnsortedList())

**Comment by lichuqiang** [comment]
@jiayingz @ScorpioCPH done.

**Comment by lichuqiang** [comment]
Seems version in open-api got updated to 1.10 and something went wrong in CI.

**Review by ScorpioCPH — COMMENTED** [review]
lgtm

**Comment by RenaudWasTaken** [comment]
In another PR (#52616) I'm removing this test to replace it other more focused test (and the architecture changes a lot too).

I'm not sure it's worth the effort to merge this one and then remove the code (and have to go through the rebase :D).
Maybe you could look at the tests on that PR and give your opinion on what tests are missing ?

**Comment by lichuqiang** [comment]
@RenaudWasTaken It's mainly for fixing test case to cover the scenario in #55179 , as commented [here](https://github.com/kubernetes/kubernetes/pull/52149#issuecomment-344801474) and [here](https://github.com/kubernetes/kubernetes/pull/52149#issuecomment-344825266).

Would look into #52616 later.

**Comment by RenaudWasTaken** [comment]
@lichuqiang I'm not sure it's worth fixing a test that we're planning to remove and replace since the architecture is changing anyways :)

**Comment by tengqm** [comment]
+1 on stop patching the current architecture. It is frustrating to rebase and rebase and rebase.

**Comment by lichuqiang** [comment]
@RenaudWasTaken SGTM, would close the PR and revisit the part if necessary.

**Pull request closed without merging**